Entities/Turret.cpp

```diff
@@ -86,6 +121,18 @@ namespace RTE {
 	void Turret::Draw(BITMAP *pTargetBitmap, const Vector &targetPos, DrawMode mode, bool onlyPhysical) const {
 		Attachable::Draw(pTargetBitmap, targetPos, mode, onlyPhysical);
 		//TODO replace this with a relative draw order property or something that lets you organize attachable drawing so it doesn't need special hardcoding crap. Use this for ahuman limbs and arm held mo if possible.
-		if (m_MountedDevice && m_MountedDevice->IsDrawnAfterParent()) { m_MountedDevice->Draw(pTargetBitmap, targetPos, mode, onlyPhysical); }
+		for (const std::unique_ptr<HeldDevice> &mountedDevice : m_MountedDevices) {
+			if (mountedDevice->IsDrawnAfterParent()) { mountedDevice->Draw(pTargetBitmap, targetPos, mode, onlyPhysical); }
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void Turret::RemoveMountedDevice(const HeldDevice *mountedDeviceToRemove) {
+		std::vector<std::unique_ptr<HeldDevice>>::iterator mountedDeviceIterator = std::find_if(m_MountedDevices.begin(), m_MountedDevices.end(), [&mountedDeviceToRemove](const std::unique_ptr<HeldDevice> &mountedDevice) {
+			return mountedDevice.get() == mountedDeviceToRemove;
+		});
+		(*mountedDeviceIterator).release();
```

Comment:
What's happening here?
You're releasing the raw pointer into the void and leaking it.
If you need to pass ownership elsewhere you need to return the pointer.

Reply:
Read the method comment :ju:

Reply:
The comment doesn't make much sense.
When you construct the unique_ptr the object is `new`d into existence. When you release it it doesn't magically become a local one that will be destructed when going out of scope.
You still need to dispose of it properly.
So I don't really know what you tried to do here, pass ownership or delete.
If it's delete then just `reset()` it.

Reply:
I'll admit, this one is complicated, because of the need to maintain this pointer in two places, one of which uses unique_ptr, and the other of which uses raw pointers. I might try and see if I can do it less weirdly.

Lemme try to explain the process flow:

1. You call `Turret::AddMountedDevice(HeldDevice *newMountedDevice)`. This adds the device to `m_MountedDevices` and calls `AddAttachable` on it, which adds it to `m_Attachables`. It then adds a `Hardcoded Attachable Remover` for the device, based on its unique ID. Said remover calls the weirdo `RemoveMountedDevice` method.

2. The only way to remove this device you added is with `MOSR::RemoveAttachable`, `RemoveMountedDevice` is private and should probably not be called in any other cases. `RemoveAttachable` removes the device from `m_Attachables` and cleans it up, then looks inside the vector of `Hardcoded Attachable Removers` to see if there's a remover to be called. If there is, it calls that remover, which is `RemoveMountedDevice`. Remember, that it still has a valid pointer to the device being removed when it calls that.

3. `RemoveMountedDevice` now does what it says in the comment - it removes the device from the `m_MountedDevices` vector, and it releases, but does NOT delete, the pointer. This is important, because `RemoveAttachable` is still in charge of dealing with that pointer, which generally means passing it to `MovableMan`. If the pointer was reset, it'd be a junk object, that couldn't go to `MovableMan`, but it has to be released from here in order for ownership to be transferrable.

Reply:
Bruh that is incredibly fucky.
So you're telling me that you have a unique_ptr that is basically not the owner of its raw ptr.
Why did you even use a unique here if that's the case? That just complicates things in a way that makes absolutely no sense.
Bjarne would lose his remaining hair if he saw this.

Reply:
Nahh, it definitely does own the raw pointer. This happens when it has to give up ownership of the raw pointer.
Though you're right, things will probably be a little less weird looking if I use raw pointers (but still owning) for this, since there'll be no need to release before removing from the list.

Reply:
Yeah you definitely should just use raw ptrs here.
Looks like it's just a matter of changing `m_MountedDevices` to be a vector of raw ptrs instead of uniques.

Reply:
Oh yeah, looks like I have a todo for how I thought to do it in Turret.h
> //TODO I think things would be cleaner if this (and all hardcoded attachable pointers) used weak_ptrs. It would solve some weird ownership stuff, particularly with this. However, for that to be possible, m_Attachables has to be shared_ptrs though.

Anyway, I'll swap it to raw pointers